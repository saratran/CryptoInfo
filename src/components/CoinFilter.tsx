import {
  Autocomplete,
  autocompleteClasses,
  Button,
  Paper,
  Popper,
  styled,
  TextField,
  Typography,
  useMediaQuery,
  useTheme,
} from "@mui/material";
import { makeStyles } from "@mui/styles";
import React, { FC, useState } from "react";
import { ListChildComponentProps, VariableSizeList } from "react-window";
import "./styles/NavBar.css";

const useStyles = makeStyles({
  root: {
    padding: 20,
  },
  autoComplete: {
    maxWidth: 800,
  },
  applyButton: {
    marginTop: 20,
  },
});

interface CoinFilterProp {
  coins: CoinSimple[];
  isLoadingCoins: boolean;
  handleFilterApply: (coins: Array<string>) => void;
}

const CoinFilter: FC<CoinFilterProp> = ({
  coins,
  isLoadingCoins,
  handleFilterApply,
}) => {
  const classes = useStyles();
  const [coinFilters, setCoinFilters] = useState<Array<string>>([]);
  return (
    <Paper className={classes.root}>
      <Autocomplete
        className={classes.autoComplete}
        multiple
        id="tags-standard"
        options={coins}
        filterSelectedOptions
        getOptionLabel={(option) => option.name}
        isOptionEqualToValue={(option, value) => option.id === value.id}
        onChange={(event: any, newValue) => {
          if (newValue) setCoinFilters(newValue.map((value) => value.id));
        }}
        loading={isLoadingCoins}
        disableListWrap
        PopperComponent={StyledPopper}
        ListboxComponent={ListboxComponent}
        renderOption={(props, option) => [props, option]}
        renderGroup={(params) => params}
        renderInput={(params) => (
          <TextField
            {...params}
            variant="standard"
            label="Coin names"
            placeholder="e.g. bitcoin"
          />
        )}
      />
      <Button
        className={classes.applyButton}
        variant="contained"
        color="primary"
        onClick={() => handleFilterApply(coinFilters)}
      >
        Apply
      </Button>
    </Paper>
  );
};

const LISTBOX_PADDING = 8; // px

function renderRow(props: ListChildComponentProps) {
  const { data, index, style } = props;
  const dataSet = data[index];
  const inlineStyle = {
    ...style,
    top: (style.top as number) + LISTBOX_PADDING,
  };

  return (
    <Typography component="li" {...dataSet[0]} noWrap style={inlineStyle}>
      {dataSet[1].name}
    </Typography>
  );
}

const OuterElementContext = React.createContext({});

const OuterElementType = React.forwardRef<HTMLDivElement>((props, ref) => {
  const outerProps = React.useContext(OuterElementContext);
  return <div ref={ref} {...props} {...outerProps} />;
});

function useResetCache(data: any) {
  const ref = React.useRef<VariableSizeList>(null);
  React.useEffect(() => {
    if (ref.current != null) {
      ref.current.resetAfterIndex(0, true);
    }
  }, [data]);
  return ref;
}

// Adapter for react-window
const ListboxComponent = React.forwardRef<
  HTMLDivElement,
  React.HTMLAttributes<HTMLElement>
>(function ListboxComponent(props, ref) {
  const { children, ...other } = props;
  const itemData: React.ReactChild[] = [];
  (children as React.ReactChild[]).forEach(
    (item: React.ReactChild & { children?: React.ReactChild[] }) => {
      itemData.push(item);
      itemData.push(...(item.children || []));
    }
  );

  const theme = useTheme();
  const smUp = useMediaQuery(theme.breakpoints.up("sm"), {
    noSsr: true,
  });
  const itemCount = itemData.length;
  const itemSize = smUp ? 36 : 48;

  const getChildSize = (child: React.ReactChild) => {
    if (child.hasOwnProperty("group")) {
      return 48;
    }

    return itemSize;
  };

  const getHeight = () => {
    if (itemCount > 8) {
      return 8 * itemSize;
    }
    return itemData.map(getChildSize).reduce((a, b) => a + b, 0);
  };

  const gridRef = useResetCache(itemCount);

  return (
    <div ref={ref}>
      <OuterElementContext.Provider value={other}>
        <VariableSizeList
          itemData={itemData}
          height={getHeight() + 2 * LISTBOX_PADDING}
          width="100%"
          ref={gridRef}
          outerElementType={OuterElementType}
          innerElementType="ul"
          itemSize={(index) => getChildSize(itemData[index])}
          overscanCount={5}
          itemCount={itemCount}
        >
          {renderRow}
        </VariableSizeList>
      </OuterElementContext.Provider>
    </div>
  );
});

const StyledPopper = styled(Popper)({
  [`& .${autocompleteClasses.listbox}`]: {
    boxSizing: "border-box",
    "& ul": {
      padding: 0,
      margin: 0,
    },
  },
});

export default CoinFilter;
