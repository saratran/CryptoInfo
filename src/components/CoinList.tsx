import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@mui/material";
import { makeStyles } from "@mui/styles";
import React, { FC } from "react";
import "./styles/NavBar.css";

const useStyles = makeStyles({
  root: {
    marginTop: 20,
    marginBottom: 20,
  },
  tableContainer: {},
  table: {},
  list: {},
  thead: {},
  tbody: {},
  row: {
    height: 40,
  },
  headerRow: {},
  headerCell: {
    fontWeight: "bold",
  },
  cell: {},
  expandingCell: {},
  column: {},
});

interface CoinListProp {
  coins: Array<any>;
}

const CoinList: FC<CoinListProp> = ({ coins }) => {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <TableContainer component={Paper}>
        <Table className={classes.table} stickyHeader>
          <TableHead className={classes.thead} component={Paper}>
            <TableRow className={classes.headerRow}>
              <TableCell className={classes.headerCell} width={4}>
                Rank
              </TableCell>
              <TableCell
                className={classes.headerCell}
                align="left"
                width={200}
              >
                Name
              </TableCell>
              <TableCell className={classes.headerCell} align="left">
                Symbol
              </TableCell>
              <TableCell className={classes.headerCell} align="right">
                Current Price (USD)
              </TableCell>
              <TableCell className={classes.headerCell} align="right">
                Market Cap (USD)
              </TableCell>
              <TableCell className={classes.headerCell} align="right">
                Circulating Supply
              </TableCell>
              <TableCell className={classes.headerCell} align="right">
                %24h
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody className={classes.tbody}>
            {coins.map((item, i) => (
              <TableRow className={classes.row} key={i}>
                <TableCell
                  scope="row"
                  component="th"
                  style={{ height: "100%" }}
                  className={classes.headerCell}
                >
                  {item.market_cap_rank}
                </TableCell>
                <TableCell>
                  <img
                    src={item.image}
                    style={{ height: 20, width: 20, marginRight: 10 }}
                  />
                  {item.name}
                </TableCell>
                <TableCell align="left">{item.symbol.toUpperCase()}</TableCell>
                <TableCell align="right">
                  ${item.current_price.toLocaleString()}
                </TableCell>
                <TableCell align="right">
                  ${item.market_cap.toLocaleString()}
                </TableCell>
                <TableCell align="right">
                  {item.circulating_supply.toLocaleString()}{" "}
                  {item.symbol.toUpperCase()}
                </TableCell>
                <TableCell
                  align="right"
                  style={{
                    color:
                      item.price_change_percentage_24h < 0 ? "red" : "green",
                  }}
                >
                  {item.price_change_percentage_24h}%
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </div>
  );
};

export default CoinList;
