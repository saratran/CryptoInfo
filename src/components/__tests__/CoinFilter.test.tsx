import { render, screen, fireEvent, getByText } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import CoinList from "../CoinList";
import renderer from "react-test-renderer";
import CoinFilter from "../CoinFilter";
import userEvent from "@testing-library/user-event";

const COINS_MOCK: CoinSimple[] = [
  {
    id: "dinoswap",
    symbol: "dino",
    name: "DinoSwap",
  },
  {
    id: "dinox",
    symbol: "dnxc",
    name: "DinoX",
  },
  {
    id: "dint-token",
    symbol: "dint",
    name: "DINT Token",
  },
  {
    id: "dionpay",
    symbol: "dion",
    name: "Dionpay",
  },
  {
    id: "dipper",
    symbol: "dip",
    name: "Dipper",
  },
];

test("Renders <CoinFilter/> in DOM without errors", () => {
  const div = document.createElement("div");
  act(() => {
    ReactDOM.render(
      <CoinFilter
        coins={COINS_MOCK}
        isLoadingCoins={false}
        handleFilterApply={(coins) => {}}
      />,
      div
    );
  });
});

test("Renders <CoinFilter/> correctly against snapshot", () => {
  const tree = renderer
    .create(
      <CoinFilter
        coins={COINS_MOCK}
        isLoadingCoins={false}
        handleFilterApply={(coins) => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("<CoinFilter/> autocomplete should render list of coin names", async () => {
  render(
    <CoinFilter
      coins={COINS_MOCK}
      isLoadingCoins={false}
      handleFilterApply={(coins) => {}}
    />
  );

  const openButton = screen.getByRole("button", { name: "Open" });
  expect(openButton).toBeInTheDocument();
  userEvent.click(openButton);
  COINS_MOCK.forEach((coin) => {
    const coinName = screen.getByText(coin.name);
    expect(coinName).toBeInTheDocument();
  });
});

test("<CoinFilter/> select and apply filter", async () => {
  let selectedCoins;
  render(
    <CoinFilter
      coins={COINS_MOCK}
      isLoadingCoins={false}
      handleFilterApply={(coins) => {
        selectedCoins = coins;
      }}
    />
  );

  const textInput = screen.getByTestId("coin-names-input");
  expect(textInput).toBeInTheDocument();
  userEvent.type(textInput, "dinoswap");
  userEvent.click(screen.getByText("DinoSwap"));
  const applyButton = screen.getByText("Apply");
  userEvent.click(applyButton);
  expect(selectedCoins).toEqual(["dinoswap"]);
});
