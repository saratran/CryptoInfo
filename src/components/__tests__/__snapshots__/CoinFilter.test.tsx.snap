// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <CoinFilter/> correctly against snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 makeStyles-root-1 css-1ps6pg7-MuiPaper-root"
>
  <div
    aria-expanded={false}
    aria-owns={null}
    className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon makeStyles-autoComplete-2 css-1kpdewa-MuiAutocomplete-root"
    data-testid="coin-names-input"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="combobox"
  >
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        id="undefined-label"
      >
        Coin names
      </label>
      <div
        className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-ghsjzk-MuiInputBase-root-MuiInput-root"
        onClick={[Function]}
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-invalid={false}
          autoCapitalize="none"
          autoComplete="off"
          autoFocus={false}
          className="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused MuiAutocomplete-input MuiAutocomplete-inputFocused css-1x51dt5-MuiInputBase-input-MuiInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          placeholder="e.g. bitcoin"
          required={false}
          spellCheck="false"
          type="text"
          value=""
        />
        <div
          className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Open"
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root makeStyles-applyButton-3 css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Apply
  </button>
</div>
`;
